### PR TITLE
Makefile: use $$ in `define` blocks

### PR DESCRIPTION
**Pull request by thockin**
Make's "define" feature (macros) is subtle and it took me a long time to
convince myself this all works. In particular, we (prior to this commit)
are terribly inconsistent about the use of `$` vs `$$`.  We mostly get
away with it because the "variables" are more like "constants", but the
inconsistency trips up some things.  For example, using `$(shell)`
inside a macro will run at macro expansion time rather than when the
resulting make code is executed.

For a contrived, but concrete example, derived from our Makefile:

```
define MACRO
ifeq ($(DBG),1)
  $(warning dbg is $(DBG))
endif
endef # macro

TGTS=a b c
$(foreach pfx, $(TGTS), $(eval $(MACRO)))

default:
	@echo $@
```

yields:

```
$ make
Makefile:8: dbg is
Makefile:8: dbg is
Makefile:8: dbg is
default

$ make DBG=1
Makefile:8: dbg is 1
Makefile:8: dbg is 1
Makefile:8: dbg is 1
default
```

This is because `$(warning)` is evaluated as the macro is expanded.
Replace that with `$(shell)` and you can see how you might end up
running a bunch of things you didn't need to run.  The fix is:

```
define MACRO
ifeq ($(DBG),1)
  $$(warning dbg is $$(DBG))
endif
endef # macro

TGTS=a b c
$(foreach pfx, $(TGTS), $(eval $(MACRO)))

default:
        @echo $@
```

which yields:

```
$ make
default

$ make DBG=1
Makefile:8: dbg is 1
Makefile:8: dbg is 1
Makefile:8: dbg is 1
default
```

We COULD have only changed `$(warning)` to `$$(warning)` and left
`$(DBG)` alone, because that's a cheap expansion.  I chose NOT to do
that here because it requires brainpower to think about this all, and it
seems easier to set a simple rule: inside a `define`/`endef` block, you
always use `$$` unless you KNOW that you NEED expansion-time evaluation
(as in the `$(prefix)` in this commit, which is effectively an argument
to the macros).

/kind cleanup

```release-note
NONE
```

### CODE REVIEW

**Comment by liggitt** [comment]
🤯 
I can tag this, but I'm not sure I can review it. Is there someone who could understand it and lgtm?

**Comment by thockin** [comment]
TL;DR:

`define` is a macro.  When we expand the macro, the body of it is evaluated, variables are expanded and commands like $(warning) and $(shell) are run.  In particular, `$$` is converted to `$`.  Once that is expanded, it is evaluated.  So to pull from the commit message example:

```
define MACRO
ifeq ($(DBG),1)
  $(warning dbg is $(DBG))
endif
endef # macro

TGTS=a b c
$(foreach pfx, $(TGTS), $(eval $(MACRO)))

default:
	@echo $@
```

when run, each `$` expression is evaluated, including the $(warning), which is executed for each expansion and the output used as replacement.  $(warning) has no output, so there's no replacement text.  This yields something like (comments added)

```
TGTS=a b c
ifeq (,1)  # pfx=a
endif
ifeq (,1)  # pfx=b
endif
ifeq (,1)  # pfx=c
endif

default:
	@echo $@
```

The main point here is that the $(warning) runs even when DBG *is not* 1.  If we change the makefile:

```
define MACRO
ifeq ($(DBG),1)
  $$(warning dbg is $$(DBG))
endif
endef # macro

TGTS=a b c
$(foreach pfx, $(TGTS), $(eval $(MACRO)))

default:
        @echo $@
```

(note the `$$(warning)`)

When that is evaluated, `$$` -> `$` and the result is something like:

```
TGTS=a b c
ifeq (,1)  # pfx=a
  $(warning dbg is $(DBG))
endif
ifeq (,1)  # pfx=b
  $(warning dbg is $(DBG))
endif
ifeq (,1)  # pfx=c
  $(warning dbg is $(DBG))
endif

default:
	@echo $@
```

Now those warnings will only fire when DBG is 1.

Now change those $(warning) to some non-trivial $(shell) (e.g. a find operation) and you can see how previously we'd execute the find even when we didn't need to.

**Comment by dims** [comment]
whew! that was something. thanks @thockin 

/approve
/lgtm
